doc/dev/error-categorization.md

```diff
@@ -0,0 +1,173 @@
+such as wrong plugin names, missing backends, initialization errors,
+misconfiguration of Elektra etc. Installation errors might also be non-Elektra
+specific but also from dependent library/applications such as gpg. Also plugin
+configuration errors belong to `Installation Errors` as this happens during
+mounting. Users will have to reconfigure, reinstall, recompile (with other
+settings) Elektra in order to get rid of this error or fix the installation of
+the corresponding library/application.
+
+#### Logical ("C01300")
+
+`Logical Errors` is a branch category in which you indicate a bug in Elektra
+such as internal errors, assertion failures, not implemented features or errors
+which should not happen such as going into a `default` branch when you are
+assured that all cases are covered. Usually such errors come with a message to
+report such failures to Elektra's bugtracker. Applications cannot handle such
+errors themselves.
+
+##### Internal ("C01310")
+
+`Internal Errors` are such errors which indicate a flaw or bug in the internal
+logic of Elektra. Examples are going into a `default` branch which you do never
+expect to happen. This category might be used in the future to automatically
+issue a bug to our bugtracker. As of now if you have to use this error please
+add a message indicating that this bug should be reported.
+
+##### Interface ("C01320")
+
+`Interface Errors` indicate a wrong usage of Elektra's API. Compared to
+`internal` errors this category does not indicate a bug but a misuse. An example
+would be to pass a NULL pointer to `kdbGet`. Also violations of the backend
+belong into this category. Compared to semantic validation errors, this category
+has its focus on detecting wrong usages of the API instead of a "retry with a
+different value" approach. Also validation errors focus on specifications of
+configurations whereas this category tries to handle specifications for APIs.
+Users should investigate the concrete reason and might use a different, more
+appropriate method or change their passed values before retrying.
+
+##### Plugin Misbehavior ("C01330")
+
+`Plugin Misbehavior Errors` indicate that a plugin does not behave in an
+intended way. Unrecognized commands, unknown return codes, plugin creation
+errors, etc. belong to this category. Also uncaught exceptions belong here
+because Elektra expects all exceptions to be caught. For wrong plugin versions
+please use `Installation` errors. Users can try to remount, recompile the plugin
+under different options or use a different plugin (e.g., switching to `mini`
+instead of `ini`).
+
+### Conflicting State ("C02000")
+
+`Conflicting State Errors` are errors where the current state is incompatible
+with the attempted operation. These kind of errors are usually in resolver
+plugins when the state of the file has changed without the system knowing.
+Examples are the need for calling `kdbGet` before `kdbSet`. Another example
```

Comment:
`kdbGet` before `kdbSet` is not an example. It is what this error tells you to do.